### PR TITLE
Add informer-based controller

### PR DESCRIPTION
**Pull request by pmorie**
Create a new informer-based controller and a new binary to run it, along with accompanying config structs and all the trimmings.

Depends on #340.

### CODE REVIEW

**Comment by krancour** [comment]
Note that since #340 was merged, 7ae91683cf88ca8e79ae48e98d83b9c2129283d5 doesn't need to be included in this PR.

**Comment by krancour** [comment]
Also wanted to capture here that @pmorie and I discussed offline (some time ago) and agreed the package name (`openservicebroker`) needs to change, as it currently hints that there's something OSB-specific about this controller, but really the controller doesn't know anything about OSB.

File: cmd/controller-manager/app/options/options.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// The controller is responsible for running control loops that reconcile
+// the state of service catalog API resources with service brokers, service
+// classes, service instances, and service bindings.
+
+package options
+
+import (
+	"github.com/kubernetes-incubator/service-catalog/pkg/apis/componentconfig"
+	"github.com/spf13/pflag"
+	k8scomponentconfig "k8s.io/kubernetes/pkg/apis/componentconfig"
```
Comment (pmorie) [inline cmd/controller-manager/app/options/options.go]: oops, used the wrong package.

File: pkg/apis/componentconfig/types.go [hunk 2]
```diff
@@ -30,3 +30,33 @@ type ControllerConfiguration struct {
 	// information.
 	KubeconfigPath string
 }
+
+// ControllerManagerConfiguration encapsulates configuration for the
+// controller manager.
+type ControllerManagerConfiguration struct {
+	// Address is the IP address to serve on (set to 0.0.0.0 for all interfaces).
+	Address string
+	// Port is the port that the controller's http service runs on.
```
Comment (krancour) [inline pkg/apis/componentconfig/types.go]: Nit: s/controller's/controller manager's/

File: cmd/controller-manager/app/controller.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
+/*
```
Comment (krancour) [inline cmd/controller-manager/app/controller.go]: Rename this file to `controller_manager.go`?

File: cmd/controller-manager/app/controller.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+// parameters.
+func NewControllerManagerCommand() *cobra.Command {
+	s := options.NewControllerManagerServer()
+	s.AddFlags(pflag.CommandLine)
+	cmd := &cobra.Command{
+		Use: "controller-manager",
+		Long: `The service-catalog controller manager is a daemon that embeds
+the core control loops shipped with the service catalog.`,
+		Run: func(cmd *cobra.Command, args []string) {
+		},
```
Comment (arschles) [inline cmd/controller-manager/app/controller.go]: this does nothing?

File: pkg/controller/client_builder.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	"k8s.io/kubernetes/pkg/client/restclient"
+
+	"github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset"
+
+	"github.com/golang/glog"
+)
+
+// ClientBuilder allow syou to get clients and configs for controllers
```
Comment (arschles) [inline pkg/controller/client_builder.go]: typo nit: "allow syo" => "allows you"

File: pkg/controller/wip/controller.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+type controller struct {
+	kubeClient           k8sclientset.Interface
+	serviceCatalogClient servicecatalogclientset.Interface
+	brokerInformer       cache.SharedInformer
+	serviceClassInformer cache.SharedInformer
+	instanceInformer     cache.SharedInformer
+	bindingInformer      cache.SharedInformer
+}
+
+var _ Controller = &controller{}
```
Comment (arschles) [inline pkg/controller/wip/controller.go]: This is unnecessary. The above `NewController` function will ensure that `controller` implements `Controller`

**Comment by krancour** [comment]
@pmorie should there be some chart updates in here also so that the new controller-manager gets deployed instead of the old controller?

**Comment by pmorie** [comment]
@krancour 

> should there be some chart updates in here also so that the new controller-manager gets deployed instead of the old controller?

Good question.  I would like to just add in the controller at this point and establish that it's working before updating things to use it.  Updating charts is something I'd like to coordinate with people about before starting on.

**Comment by krancour** [comment]
@pmorie so everyone who wants to verify this works needs to go modify the chart themselves? It seems it would be more efficient to bundle it in here.

**Comment by pmorie** [comment]
@krancour I'm going to be working on testing this tomorrow -- let's talk about chart updates then, if you have some time.

**Comment by pmorie** [comment]
I've addressed the nits that have been left so far.  This is currently working correctly for me using the `catalog-wip` chart using the following helm command to install:

```
helm install --namespace=service-catalog --set version=latest,debug=1,nodePort=30000,imagePullPolicy=IfNotPresent,insecure=true ~/code/service-catalog/deploy/wip-catalog/ -n sc
```

I suggest that we merge this and being porting in business logic and address nits later in follow-ups.

**Comment by arschles** [comment]
I am making the administrative decision to merge this now, and follow @pmorie's suggestion to address nits and business logic porting in follow-ups.

**Pull request merged**